datafusion/core/src/datasource/physical_plan/parquet/row_groups.rs

```diff
@@ -168,92 +174,94 @@ pub(crate) async fn prune_row_groups_by_bloom_filters<
-    }
 
-    fn prune(&self, column_sbbf: &HashMap<String, Sbbf>) -> bool {
-        Self::prune_expr_with_bloom_filter(self.predicate_expr.as_ref(), column_sbbf)
+        Ok(Self {
+            schema,
+            predicate_expr: expr.clone(),
+            required_columns: columns.into_iter().collect(),
+        })
     }
 
     /// filter the expr with bloom filter return true if the expr can be pruned
-    fn prune_expr_with_bloom_filter(
-        expr: Option<&phys_expr::BinaryExpr>,
-        column_sbbf: &HashMap<String, Sbbf>,
-    ) -> bool {
-        if expr.is_none() {
-            return false;
-        }
-        let expr = expr.unwrap();
-        match expr.op() {
-            Operator::And => {
-                let left = Self::prune_expr_with_bloom_filter(
-                    expr.left().as_any().downcast_ref::<phys_expr::BinaryExpr>(),
-                    column_sbbf,
-                );
-                let right = Self::prune_expr_with_bloom_filter(
-                    expr.right()
-                        .as_any()
-                        .downcast_ref::<phys_expr::BinaryExpr>(),
-                    column_sbbf,
-                );
-                left || right
-            }
-            Operator::Or => {
-                let left = Self::prune_expr_with_bloom_filter(
-                    expr.left().as_any().downcast_ref::<phys_expr::BinaryExpr>(),
-                    column_sbbf,
-                );
-                let right = Self::prune_expr_with_bloom_filter(
-                    expr.right()
-                        .as_any()
-                        .downcast_ref::<phys_expr::BinaryExpr>(),
-                    column_sbbf,
-                );
-                left && right
-            }
-            Operator::Eq => {
-                if let Some((col, val)) = Self::check_expr_is_col_equal_const(expr) {
-                    if let Some(sbbf) = column_sbbf.get(col.name()) {
-                        match val {
-                            ScalarValue::Utf8(Some(v)) => !sbbf.check(&v.as_str()),
-                            ScalarValue::Boolean(Some(v)) => !sbbf.check(&v),
-                            ScalarValue::Float64(Some(v)) => !sbbf.check(&v),
-                            ScalarValue::Float32(Some(v)) => !sbbf.check(&v),
-                            ScalarValue::Int64(Some(v)) => !sbbf.check(&v),
-                            ScalarValue::Int32(Some(v)) => !sbbf.check(&v),
-                            ScalarValue::Int16(Some(v)) => !sbbf.check(&v),
-                            ScalarValue::Int8(Some(v)) => !sbbf.check(&v),
-                            _ => false,
-                        }
-                    } else {
-                        false
-                    }
+    fn prune(&self, column_sbbf: &HashMap<String, Sbbf>) -> bool {
+        // rewrite any <col = literal> exprs to to `false` if we can prove they
+        // are always false using the bloom filter.
+        let rewritten = self
```

Comment:
I am quite pleased with this logic. What it does is to rewrite the predicate `String@0 = Hello_Not_Exists AND 1 = 1` is rewritten to `false AND false` and the interval analysis determines that expression is always `false`:

```
Evaluating. self.predicate_expr: String@0 = Hello_Not_Exists AND 1 = 1
   rewritten false AND 1 = 1
   interval result: [false, false]
lower: IntervalBound { value: Boolean(false), open: false }
upper: IntervalBound { value: Boolean(false), open: false }
```

However it seems I need to update the interval analysis so that `OR` is handled  as  `OR` gives a "unknown" for the bounds:

```
Evaluating. self.predicate_expr: String@0 = Hello_Not_Exists OR String@0 = Hello_Not_Exists2
   rewritten false OR false
   interval result: (NULL, NULL)
lower: IntervalBound { value: NULL, open: true }
upper: IntervalBound { value: NULL, open: true }
```

Reply:
Added support for OR here: https://github.com/apache/arrow-datafusion/pull/7884